### ISSUE TITLE
MariaDB destroys JSON when a " is present inside any of the fields

### ISSUE THREAD

**Issue by MySidesTheyAreGone** [lukechilds/keyv-mysql#15]
This is super-strange because it doesn't happen at all with Postgres or SQLite. Since the other bug was much more restricted than I thought, here's a quick test you can run:

https://gist.github.com/MySidesTheyAreGone/d2ce13b1f401bb5f62782ddcc0c9015e

I ran it straight from Keyv's directory, if you don't then adjust the "require" accordingly. Adjust mysql's username/pwd/address. If it doesn't fail for you too then I don't know what's happening anymore.

The test object is a fragment of what Reddit's API returns for self posts.

Adding a couple of `console.log`s in keyv-sql to make things more obvious - print any SQL and print the values as they are when fetched from the DB, before they are fed to JSONB.parse - I got this:

```
mysql://root:devlocal@localhost:3306/test

[keyv-sql] Running query:
REPLACE INTO `test` (`key`, `value`) VALUES ('test:testkey', '{"value":{"reddit:subreddit":"textventures","reddit:selftext_html":"<!-- SC_OFF --><div class=\"md\"><p>Hello fellow redditors</p>[...]</div><!-- SC_ON -->","key":"reddit:t3_6tv00l"},"expires":null}')

[keyv-sql] Running query:
SELECT `test`.* FROM `test` WHERE (`test`.`key` = 'test:testkey')

[keyv-sql] Got value:
{"value":{"reddit:subreddit":"textventures","reddit:selftext_html":"<!-- SC_OFF --><div class="md"><p>Hello fellow redditors</p>[...]</div><!-- SC_ON -->","key":"reddit:t3_6tv00l"},"expires":null}

Error: While getting back key "testkey" from storage, expecting value {"reddit:subreddit":"textventures","reddit:selftext_html":"<!-- SC_OFF --><div class=\"md\"><p>Hello fellow redditors</p>[...]</div><!-- SC_ON -->","key":"reddit:t3_6tv00l"}
    at setget (/home/mysides/node/keyv/test.js:37:15)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)

SyntaxError: Unexpected token m in JSON at position 95
    at JSON.parse (<anonymous>)
    at Object.exports.parse (/home/mysides/node/keyv/node_modules/json-buffer/index.js:49:15)
    at Promise.resolve.then.then.data (/home/mysides/node/keyv/src/index.js:54:47)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
```

If you look closely at the `[keyv-sql] Got value:` bit, the quotes are no longer escaped (look ad `"md"`). That's why JSONB.parse complains about an unexpected character `m` - it's the `m` of `md`. Note that the error happens in Keyv, not in the test script.

True enough, when I look at the contents of the table the slashes have disappeared; and they *always* disappear no matter how the `REPLACE INTO` query is run - from Node or from DBeaver. Clearly, MariaDB is removing one "round" of escaping from `value`. I believe we need to double all the slashes so that the value is stored correctly; if I do that, it works fine.

If you want the absolute smallest test case, replace line 48 of the gist with:
```
await setget(keyv, 'testkey', {a: '"'})
```
You'll end up with this inside the `value` column:

`{"value":{"a":"""},"expires":null}`

...which is invalid JSON.

This is SO strange.

Arch Linux, MariaDB 10.1.26

I can try MySQL in about 8-9 hours or so.

**Comment by MySidesTheyAreGone** [comment]
So, SQLfiddle is a thing apparently.

http://sqlfiddle.com/#!9/b28da2/83/0

http://sqlfiddle.com/#!9/b28da2/84

http://sqlfiddle.com/#!9/b28da2/85

¯\\_(ツ)_/¯  <- had to escape his right arm

**Comment by lukechilds** [comment]
Thanks for reporting, you're on fire 🔥

Can repro with `{a: '"'}`.

Also affect MySQL: `Ver 5.7.19 for Linux on x86_64 (MySQL Community Server (GPL))`

**Comment by MySidesTheyAreGone** [comment]
WE CAN DO THIS 💪💪💪💪💪

**Comment by lukechilds** [comment]
Pretty fugly but this seems to do the trick inside `.set()`:

```js
if (this.opts.dialect === 'mysql') {
  value = value.replace(/\\/g, "\\\\");
}
```

**Comment by MySidesTheyAreGone** [comment]
Yep, it works now!

**Comment by lukechilds** [comment]
Yep, just published `@keyv/mysql` v1.0.5 with the fix. Thanks for your help!

**Comment by MySidesTheyAreGone** [comment]
https://media.giphy.com/media/9o67upvAnOqRy/giphy.gif
